### PR TITLE
PyTorch Ignite example

### PR DESCRIPTION
**Pull request by scarpma**
@rusty1s I created the pull request connected to issue #4454 . This is my example. We can work together to adjust the content, the style and the amount of comments. For now it includes a lot of things. Maybe tensorboard logging could be removed in order to make it more simple.

Let me know if you like it.

And thanks for the explanation about how to make the pull request.

### CODE REVIEW

File: examples/pytorch_ignite/gin.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        test_evaluator,
+        event_name=ignite.engine.Events.EPOCH_COMPLETED,
+        tag="test",
+        metric_names='all',
+        global_step_transform=ignite.handlers.global_step_from_engine(trainer),
+    )
+    tb_logger.close()
+
+    # run the trainer
+    trainer.run(loader_trn, max_epochs=50)
```
Comment (scarpma) [inline examples/pytorch_ignite/gin.py:210]: The perfomance of the model are pretty bad. Don't know if this is normal, being just an example (I didn't run the original example with pytorch lighting).

File: examples/pytorch_ignite/gin.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    def log_validation_metrics(trainer):
+        val_evaluator.run(loader_val)
+        metrics = val_evaluator.state.metrics
+        print(f"{'Validation Results - Epoch: ':30}{trainer.state.epoch} ",
+              end='')
+        for key, value in metrics.items():
+            print(f"{key}: {value:.2f} ", end='')
+        print()
+
+    test_evaluator = ignite.engine.create_supervised_evaluator(
```
Comment (scarpma) [inline examples/pytorch_ignite/gin.py:130]: I have to create three different evaluators (even if they are equal) because, after creation, I attach different functions for personalized logging to each one. Don't know if there's a better solution to this.

**Review by scarpma — COMMENTED** [review]
Only added some comments. Don't know how to procede.

File: examples/pytorch_ignite/gin.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+    dataset = dataset.shuffle()
+    test_dataset = dataset[:len(dataset) // 10]
+    val_dataset = dataset[len(dataset) // 10:2 * len(dataset) // 10]
+    train_dataset = dataset[2 * len(dataset) // 10:]
+    loader_trn = pyg.data.DataLoader(
+        dataset=train_dataset,
+        batch_size=batch_size,
+        shuffle=True,
+        num_workers=1,
```
Comment (rusty1s) [inline examples/pytorch_ignite/gin.py:51]:
```suggestion
        num_workers=1,
```
I think we either want to increase `num_workers` or set it to `0`. In my benchmarks, `num_workers=1` performs quite poorly.

File: examples/pytorch_ignite/gin.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+    model = Model(dataset.num_node_features, dataset.num_classes).to(device)
+    optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
+    loss = torch.nn.functional.cross_entropy
+    metrics = {'acc': ignite.metrics.Accuracy()}
+
+    def prepare_batch_fn(batch, device, non_blocking):
+        # return a tuple of (x, y) that can be directly runned as
+        # `loss_fn(model(x), y)`
+        # dato che ignite non dà argomenti multipli a model, dovremo
```
Comment (rusty1s) [inline examples/pytorch_ignite/gin.py:77]: I cannot understand this comment :(

File: examples/pytorch_ignite/gin.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        # avere modelli che estraggono dentro forward() i dati di input
+        # da `data`
+        return (batch.to(device, non_blocking=non_blocking),
+                batch.y.to(device, non_blocking=non_blocking))
+
+    trainer = ignite.engine.create_supervised_trainer(
+        model=model,
+        optimizer=optimizer,
+        loss_fn=loss,
```
Comment (rusty1s) [inline examples/pytorch_ignite/gin.py:86]:
```suggestion
        loss_fn=F.cross_entropy,
```

File: examples/pytorch_ignite/gin.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        device=device,
+        prepare_batch=prepare_batch_fn,
+        output_transform=lambda x, y, y_pred, loss: loss.item(),
```
Comment (rusty1s) [inline examples/pytorch_ignite/gin.py:89]: I'm not familiar with Ignite. Is the `loss.item()` here correct (since it will detach the loss value from the computation graph)?

File: examples/pytorch_ignite/gin.py (hunk 2 above)
Comment (rusty1s) [reply examples/pytorch_ignite/gin.py:130]: Can't we generate a single function `log_metrics` and call it within `log_training_metrics`, etc?

File: examples/pytorch_ignite/gin.py (hunk 1 above)
Comment (rusty1s) [reply examples/pytorch_ignite/gin.py:210]: What does that mean? What's the final performance of the model?

Comment (scarpma) [reply examples/pytorch_ignite/gin.py:210]: For the training set like 0.80 - 0.85 accuracy, starting at epoch 0 from 0.6-0.7. For the validation and test the accuracy does not increase during training

File: examples/pytorch_ignite/gin.py (hunk 4 above)
Comment (scarpma) [reply examples/pytorch_ignite/gin.py:77]: ops, sorry. I'm gonna change this

File: examples/pytorch_ignite/gin.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+        batch_size=batch_size,
+        shuffle=True,
+        num_workers=1,
+        pin_memory=torch.cuda.is_available(),
+    )
+    loader_val = pyg.data.DataLoader(
+        dataset=val_dataset,
+        batch_size=batch_size,
+        shuffle=False,
+        num_workers=1,
```
Comment (scarpma) [inline examples/pytorch_ignite/gin.py:58]:
```suggestion
        num_workers=0,
```

File: examples/pytorch_ignite/gin.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        pin_memory=torch.cuda.is_available(),
+    )
+    loader_test = pyg.data.DataLoader(
+        dataset=test_dataset,
+        batch_size=batch_size,
+        shuffle=False,
+        num_workers=1,
```
Comment (scarpma) [inline examples/pytorch_ignite/gin.py:65]:
```suggestion
        num_workers=0,
```

File: examples/pytorch_ignite/gin.py (hunk 3 above)
Comment (scarpma) [inline examples/pytorch_ignite/gin.py:51]:
```suggestion
        num_workers=0,
```

File: examples/pytorch_ignite/gin.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+        pin_memory=torch.cuda.is_available(),
+    )
+
+    model = Model(dataset.num_node_features, dataset.num_classes).to(device)
+    optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
+    loss = torch.nn.functional.cross_entropy
```
Comment (scarpma) [inline examples/pytorch_ignite/gin.py:71]:
```suggestion

```

File: examples/pytorch_ignite/gin.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+        # dato che ignite non dà argomenti multipli a model, dovremo
+        # avere modelli che estraggono dentro forward() i dati di input
+        # da `data`
+        return (batch.to(device, non_blocking=non_blocking),
+                batch.y.to(device, non_blocking=non_blocking))
+
+    trainer = ignite.engine.create_supervised_trainer(
+        model=model,
+        optimizer=optimizer,
+        loss_fn=F.cross_entropy,
```
Comment (scarpma) [inline examples/pytorch_ignite/gin.py:86]:
```suggestion
        loss_fn=torch.nn.functional.cross_entropy,
```

**Comment by scarpma** [comment]
Any news on this ?

**Comment by rusty1s** [comment]
Will try to merge in the upcoming week. Sorry for the delay!

**Comment by scarpma** [comment]
> Will try to merge in the upcoming week. Sorry for the delay!

Ok perfect, don't worry. I was just checking in the case this was forgotten.

**Comment by rusty1s** [comment]
Sorry again for the delay. I currently do not have access to a GPU and still want to run it before merging.

**Comment by scarpma** [comment]
Ok, no worries. Let me know !

Il ven 20 mag 2022, 05:53 Matthias Fey ***@***.***> ha
scritto:

**Review by rusty1s — APPROVED** [review]
Thank you!
